crypto/params.c

```diff
@@ -0,0 +1,369 @@
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT_SET(to, from) \
+    case sizeof(to): \
+        SET_RETURN_SIZE(p, sizeof(to)); \
+        if (sizeof(to) >= sizeof(from)) { \
+            *(to *)p->buffer = (to)val; \
+            return 1; \
+        } \
+        break
+
+#define PARAM_INT(name, type, pre) \
+    int OSSL_PARAM_get_##name(const OSSL_PARAM *p, type *val) \
+    { \
+        if (val == NULL \
+            || p == NULL \
+            || (p->data_type != OSSL_PARAM_INTEGER \
+                && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER)) \
+            return 0; \
+        switch (p->buffer_size) { \
+        PARAM_INT_GET(type, pre ## 32_t); \
+        PARAM_INT_GET(type, pre ## 64_t); \
+        } \
+        return 0; \
+    } \
+    int OSSL_PARAM_set_##name(const OSSL_PARAM *p, type val) \
+    { \
+        if (p == NULL) \
+            return 0; \
+        SET_RETURN_SIZE(p, 0); \
+        if (p->data_type == OSSL_PARAM_INTEGER \
+            || p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) \
+            switch (p->buffer_size) { \
+            PARAM_INT_SET(pre ## 32_t, type); \
+            PARAM_INT_SET(pre ## 64_t, type); \
+            } \
+        return 0; \
+    } \
+    OSSL_PARAM OSSL_PARAM_construct_##name(const char *key, type *buf, \
+                                           size_t *rsize) \
+    { \
+        return ossl_param_construct(key, #pre[0] == 'u' \
+                                         ? OSSL_PARAM_UNSIGNED_INTEGER \
+                                         : OSSL_PARAM_INTEGER, \
+                                    buf, sizeof(type), rsize); \
+    }
+
+PARAM_INT(int, int, int)
+PARAM_INT(uint, unsigned int, uint)
+PARAM_INT(long, long int, int)
+PARAM_INT(ulong, unsigned long int, uint)
+PARAM_INT(int32, int32_t, int)
+PARAM_INT(uint32, uint32_t, uint)
+PARAM_INT(int64, int64_t, int)
+PARAM_INT(uint64, uint64_t, uint)
+PARAM_INT(size_t, size_t, uint)
+
+OSSL_PARAM OSSL_PARAM_construct_BN(const char *key, unsigned char *buf,
+                                   size_t bsize, size_t *rsize)
+{
+    return ossl_param_construct(key, OSSL_PARAM_UNSIGNED_INTEGER,
+                                buf, bsize, rsize);
+}
+
+int OSSL_PARAM_get_BN(const OSSL_PARAM *p, BIGNUM **val)
+{
+    BIGNUM *b;
+
+    if (val == NULL || p == NULL)
+        return 0;
+
+    if (p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
+        b = BN_native2bn(p->buffer, (int)p->buffer_size, *val);
+        if (b != NULL) {
+            *val = b;
+            return 1;
+        }
+    }
+    return 0;
+}
+
+int OSSL_PARAM_set_BN(const OSSL_PARAM *p, const BIGNUM *val)
+{
+    size_t bytes;
+
+    if (p == NULL)
+        return 0;
+
+    if (val != NULL && p->data_type == OSSL_PARAM_UNSIGNED_INTEGER) {
+        bytes = (size_t)BN_num_bytes(val);
+        SET_RETURN_SIZE(p, bytes);
+        return p->buffer_size >= bytes
+               && BN_bn2nativepad(val, p->buffer, p->buffer_size) >= 0;
```

Comment:
Actually, it makes more sense to use `bytes` as length.  Think about it, the other end will use `*return_size` to determine how many buffer bytes they will use, so why waste cycles on unnecessary padding?